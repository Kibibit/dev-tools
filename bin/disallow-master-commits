#!/usr/bin/env node
const { disallowMainBranchesCommits } = require('../lib/disallow-master-commits');
disallowMainBranchesCommits();
