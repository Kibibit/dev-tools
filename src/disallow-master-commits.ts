import simpleGit from 'simple-git/promise';

import { checkIsMainBranchCheckedOut, getMainBranch } from './main-branch';

const git = simpleGit();

export async function disallowMainBranchesCommits() {
  try {
    const branchSummaryResult = await git.branch(['-vv']);
    const mainBranch = getMainBranch(branchSummaryResult);
    const isMainBranchCheckedOut =
      checkIsMainBranchCheckedOut(branchSummaryResult);

    if (isMainBranchCheckedOut) {
      console.log([
        'Should not commit directly to ',
        `${ isMainBranchCheckedOut } branch when working locally`
      ].join(''));
      process.exit(1);
    }
    process.exit(0);
  } catch (err) {
    console.error(err);
    process.exit(1);
  }
}
