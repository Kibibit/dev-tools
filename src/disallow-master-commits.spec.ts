/* eslint-disable @typescript-eslint/no-explicit-any */
import mockConsole from 'jest-mock-console';
import { mockProcessExit } from 'jest-mock-process';
import simpleGit from 'simple-git/promise';

import { disallowMainBranchesCommits } from './disallow-master-commits';

describe('Disallow Master Branch Commits', () => {
  let mockExit;
  let restoreConsole;

  beforeEach(() => {
    jest.clearAllMocks();
    mockExit = mockProcessExit();
    restoreConsole = mockConsole();
    (simpleGit as any).clearMocks();
  });

  afterEach(() => {
    mockExit.mockRestore();
    restoreConsole();
  });

  it('should allow commits if on normal branch', async () => {
    await disallowMainBranchesCommits();
    expect(mockExit).toHaveBeenCalledWith(0);
  });

  describe('should DISALLOW commits if on some main branch', () => {
    test.each([
      ['master'],
      ['main'],
      ['develop'],
      ['dev'],
      ['staging'],
      ['next'],
      ['beta'],
      ['alpha']
    ])(
      'main branch: %s',
      async (branchName) => {
        (simpleGit as any).branchReturnObj = mockMainBranch(branchName);
        await disallowMainBranchesCommits();
        expect(mockExit).toHaveBeenCalledWith(1);
        expect(console.log).toHaveBeenCalledTimes(1);
        expect((console.log as jest.Mock).mock.calls[0][0])
          .toMatchSnapshot();
      }
    );
});
});

function mockMainBranch(name = 'main', isCurrent = true) {
  return {
    all: [ name ],
    branches: {
      main: {
        current: false,
        name: name,
        commit: 'cd2ec48',
        label: '[origin/main: behind 16] ci(github): add pipelines'
      },
      'release/fix-release': {
        current: false,
        name: 'release/fix-release',
        commit: '8f64df7',
        label: '[origin/release/fix-release: gone] blah'
      }
    },
    current: isCurrent ? name : 'release/fix-release',
    detached: false
  };
}
