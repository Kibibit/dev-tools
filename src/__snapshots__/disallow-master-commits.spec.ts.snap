// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Disallow Master Branch Commits should DISALLOW commits if on some main branch main branch: alpha 1`] = `"Should not commit directly to alpha branch when working locally"`;

exports[`Disallow Master Branch Commits should DISALLOW commits if on some main branch main branch: beta 1`] = `"Should not commit directly to beta branch when working locally"`;

exports[`Disallow Master Branch Commits should DISALLOW commits if on some main branch main branch: dev 1`] = `"Should not commit directly to dev branch when working locally"`;

exports[`Disallow Master Branch Commits should DISALLOW commits if on some main branch main branch: develop 1`] = `"Should not commit directly to develop branch when working locally"`;

exports[`Disallow Master Branch Commits should DISALLOW commits if on some main branch main branch: main 1`] = `"Should not commit directly to main branch when working locally"`;

exports[`Disallow Master Branch Commits should DISALLOW commits if on some main branch main branch: master 1`] = `"Should not commit directly to master branch when working locally"`;

exports[`Disallow Master Branch Commits should DISALLOW commits if on some main branch main branch: next 1`] = `"Should not commit directly to next branch when working locally"`;

exports[`Disallow Master Branch Commits should DISALLOW commits if on some main branch main branch: staging 1`] = `"Should not commit directly to staging branch when working locally"`;
